From 1f3fd93e373de19f4f5a0f965d04747ea354e90c Mon Sep 17 00:00:00 2001
From: Devi Priya <quic_devipriy@quicinc.com>
Date: Tue, 27 Aug 2024 14:28:51 +0530
Subject: [PATCH] pcie: qcom: ipq54xx: Add support to update rate adapter regs

Add support to update the rate adapter registers for ipq54xx.
Also update the PARF_AXI_MSTR MASK to 0x3F which corresponds to 6 bits
of ADDR_BIT_INDEX (5:0)

Change-Id: Icf65bc38e982ebba51efe1153c0750c5c92af1c1
Signed-off-by: Devi Priya <quic_devipriy@quicinc.com>
---
 drivers/pci/controller/dwc/pcie-qcom.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/drivers/pci/controller/dwc/pcie-qcom.c b/drivers/pci/controller/dwc/pcie-qcom.c
index 908263ca096a..816b08b5ae03 100644
--- a/drivers/pci/controller/dwc/pcie-qcom.c
+++ b/drivers/pci/controller/dwc/pcie-qcom.c
@@ -47,7 +47,7 @@
 #define PARF_MHI_CLOCK_RESET_CTRL		0x174
 #define PARF_AXI_MSTR_WR_ADDR_HALT		0x178
 #define PARF_AXI_MSTR_WR_ADDR_HALT_V2		0x1a8
-#define PARF_AXI_MSTR_WR_ADDR_HALT_V2_MASK	0x1F
+#define PARF_AXI_MSTR_WR_ADDR_HALT_V2_MASK	0x3F
 #define PARF_Q2A_FLUSH				0x1ac
 #define PARF_LTSSM				0x1b0
 #define PARF_SID_OFFSET				0x234
@@ -302,6 +302,7 @@ struct qcom_pcie {
 	struct dentry *debugfs;
 	bool suspended;
 	uint32_t axi_wr_addr_halt;
+	uint32_t aggr_noc_rate_adap_val;
 	uint32_t domain;
 	uint32_t num_lanes;
 	int global_irq;
@@ -1220,10 +1221,14 @@ static int qcom_pcie_post_init(struct qcom_pcie *pcie)
 			pcie->parf + PARF_AXI_MSTR_WR_ADDR_HALT_V2);
 	}
 
-	if (pcie->aggr_noc != NULL && !IS_ERR(pcie->aggr_noc))
-		writel(AGGR_NOC_PCIE_1LANE_RATEADAPT_VAL, pcie->aggr_noc);
+	if (!IS_ERR_OR_NULL(pcie->aggr_noc)) {
+		if (pcie->aggr_noc_rate_adap_val)
+			writel(pcie->aggr_noc_rate_adap_val, pcie->aggr_noc);
+		else
+			writel(AGGR_NOC_PCIE_1LANE_RATEADAPT_VAL, pcie->aggr_noc);
+	}
 
-	if (pcie->system_noc != NULL && !IS_ERR(pcie->system_noc)) {
+	if (!IS_ERR_OR_NULL(pcie->system_noc)) {
 		if (pcie->num_lanes == 2)
 			writel(SYSTEM_NOC_PCIE_RATEADAPT_BYPASS, pcie->system_noc);
 	}
@@ -1779,6 +1784,9 @@ static int qcom_pcie_probe(struct platform_device *pdev)
 		}
 	}
 
+	of_property_read_u32(pdev->dev.of_node, "aggr-noc-val",
+				&pcie->aggr_noc_rate_adap_val);
+
 	of_property_read_u32(pdev->dev.of_node, "axi-halt-val",
 				&pcie->axi_wr_addr_halt);
 
-- 
2.34.1

